jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt

```diff
@@ -0,0 +1,132 @@
+import com.intellij.util.io.EnumeratorStringDescriptor
+import com.intellij.util.io.KeyDescriptor
+import org.jetbrains.yaml.YAMLLanguage
+import org.jetbrains.yaml.psi.YAMLKeyValue
+import software.aws.toolkits.jetbrains.services.cloudformation.IndexedFunction
+import software.aws.toolkits.jetbrains.services.cloudformation.IndexedResource
+import software.aws.toolkits.jetbrains.services.cloudformation.LAMBDA_FUNCTION_TYPE
+import software.aws.toolkits.jetbrains.services.cloudformation.SERVERLESS_FUNCTION_TYPE
+import software.aws.toolkits.jetbrains.services.cloudformation.yaml.YamlCloudFormationTemplate
+import java.io.DataInput
+import java.io.DataOutput
+
+class SamTemplateIndex : FileBasedIndexExtension<String, MutableList<IndexedResource>>() {
+    private val fileFilter by lazy {
+        val supportedFiles = arrayOf(YAMLLanguage.INSTANCE.associatedFileType)
+
+        object : DefaultFileTypeSpecificInputFilter(*supportedFiles) {
+            override fun acceptInput(file: VirtualFile): Boolean = file.isInLocalFileSystem
+        }
+    }
+
+    override fun getValueExternalizer(): DataExternalizer<MutableList<IndexedResource>> = object : DataExternalizer<MutableList<IndexedResource>> {
+        override fun save(dataOutput: DataOutput, value: MutableList<IndexedResource>) {
+            dataOutput.writeInt(value.size)
+            value.forEach { resource ->
+                dataOutput.writeInt(resource.indexedProperties.size)
+                resource.indexedProperties.forEach { key, value ->
+                    dataOutput.writeUTF(key)
+                    dataOutput.writeUTF(value)
+                }
+            }
+        }
+
+        override fun read(dataInput: DataInput): MutableList<IndexedResource> {
+            val size = dataInput.readInt()
+            val resources = mutableListOf<IndexedResource>()
+            repeat(size) {
+                val mutableMap: MutableMap<String, String> = mutableMapOf()
+
+                val propertySize = dataInput.readInt()
+                repeat(propertySize) {
+                    val key = dataInput.readUTF()
+                    val value = dataInput.readUTF()
+                    mutableMap[key] = value
+                }
+                resources.add(IndexedResource(mutableMap))
+            }
+            return resources
+        }
+    }
+
+    override fun getName(): ID<String, MutableList<IndexedResource>> = NAME
+
+    override fun getIndexer(): DataIndexer<String, MutableList<IndexedResource>, FileContent> = DataIndexer { fileContent ->
+        val indexedResources = mutableMapOf<String, MutableList<IndexedResource>>()
+
+        fileContent.psiFile.acceptNode(object : PsiElementVisitor() {
+            override fun visitElement(element: PsiElement?) {
+                super.visitElement(element)
+                element?.run {
+                    val parent = element.parent as? YAMLKeyValue ?: return
+                    if (parent.value != this) return
```

Comment:
what is this guarding against?

Reply:
This makes sure we only stop at the `YAMLKeyValue.value` element for this visit. The parent could have other children which we don't care.